src/librustc_ast_lowering/expr.rs

```diff
@@ -483,14 +484,44 @@ impl<'hir> LoweringContext<'_, 'hir> {
             Some(ty) => FunctionRetTy::Ty(ty),
             None => FunctionRetTy::Default(span),
         };
-        let ast_decl = FnDecl { inputs: vec![], output };
+
+        let task_context_id = self.resolver.next_node_id();
+        let task_context_hid = self.lower_node_id(task_context_id);
+        let ast_decl = FnDecl {
+            inputs: vec![Param {
+                attrs: AttrVec::new(),
+                ty: AstP(Ty {
+                    id: self.resolver.next_node_id(),
+                    kind: TyKind::Infer,
+                    span: DUMMY_SP,
+                }),
+                pat: AstP(Pat {
+                    id: task_context_id,
+                    kind: PatKind::Ident(
+                        BindingMode::ByValue(Mutability::Mut),
+                        Ident::with_dummy_span(sym::_task_context),
+                        None,
+                    ),
+                    span: DUMMY_SP,
+                }),
+                id: self.resolver.next_node_id(),
+                span: DUMMY_SP,
+                is_placeholder: false,
+            }],
+            output,
+        };
```

Comment:
I don't think we should be creating `syntax::ast` IR inside lowering just to immediately convert it to HIR. Instead, the HIR representation should be created directly. (cc @eddyb)

Reply:
That makes sense to me, but there seems to be no existing code that directly creates an `hir::FnDecl`, and `lower_fn_decl` does quite a lot of work that I don't really want to hand-roll.

Reply:
It's not that much code tbh which I think can be mostly shared by some refactorings of `lower_fn_decl` into the main parts (inputs, output, implicit_self, packaging those aforementioned up) as separate methods.

Reply:
I've pushed a commit that should address this